src/logging.rs

```diff
@@ -0,0 +1,157 @@
+use std::fs::File;
+use std::io::Write;
+use std::os::unix::io::FromRawFd;
+use std::ffi::CString;
+use std::os::unix::net::UnixDatagram;
+use std::os::unix::io::{AsRawFd, IntoRawFd};
+use nix::sys::socket::{ControlMessage, MsgFlags, sendmsg};
+use nix::sys::memfd::MemFdCreateFlag;
+use nix::sys::memfd::memfd_create;
+use nix::fcntl::*;
+
+use crate::errors::*;
+
+/// Default path of the journald AF_UNIX datagram socket.
+const SD_JOURNAL_SOCK_PATH: &str = "/run/systemd/journal/socket";
+
+/// Log priority values.
+///
+/// See `man 3 syslog`.
+#[derive(Debug)]
+#[repr(u8)]
+pub enum Priority {
+    /// system is unusable
+    Emergency = 0,
+    /// action must be taken immediately
+    Alert,
+    /// critical conditions
+    Critical,
+    /// error conditions
+    Error,
+    /// warning conditions
+    Warning,
+    /// normal, but significant, condition
+    Notice,
+    /// informational message
+    Info,
+    /// debug-level message
+    Debug,
+}
+
+impl std::convert::From<Priority> for u8 {
+    fn from(p: Priority) -> Self {
+        match p {
+            Priority::Emergency => 0,
+            Priority::Alert => 1,
+            Priority::Critical => 2,
+            Priority::Error => 3,
+            Priority::Warning => 4,
+            Priority::Notice => 5,
+            Priority::Info => 6,
+            Priority::Debug => 7,
+        }
+    }
+}
+
+#[inline(always)]
+fn is_valid_char(c: char) -> bool {
+    c.is_uppercase() || c.is_numeric() || c == '_'
+}
+
+/// The variable name must be in uppercase and consist only of characters, 
+/// numbers and underscores, and may not begin with an underscore.
+fn is_valid_field(input: &str) -> bool {
+    if input.is_empty() {
+        return false;
+    }
+
+    if !input.chars().all(is_valid_char) {
+        return false;
+    }
+
+    if input.starts_with('_') {
+        return false;
+    }
+
+    true
+}
+
+fn add_field_and_payload(data: &mut String, field: &str, payload: &str) {
+    if is_valid_field(field) {
+        let field_payload = format!("{}={}\n", field, payload);
+        data.push_str(&field_payload);
+    }
+}
+
+/// Print a message to the systemd journal with the given priority.
+pub fn sd_journal_print(priority: Priority, msg: &str) -> Result<()> {
+    let sock = UnixDatagram::unbound()?;
+    sock.connect(SD_JOURNAL_SOCK_PATH)?;
+
+    let mut data = String::new();
+    add_field_and_payload(&mut data, "PRIORITY", &(u8::from(priority)).to_string());
+    add_field_and_payload(&mut data, "MESSAGE", msg);
+
+    let res = sock.send(data.as_bytes());
```

Comment:
I think the C and Go library are using unconnected sockets and just doing writes with an explicit destination.

Do you think it would be possible to do it here as well? That is, getting rid of the `sock.connect()` and adding the address to the two send operations?

(If it starts getting to unwieldy, I'm fine with merging this initial logic and maybe following up in a later PR)

Reply:
It looks possible, the only issue will be the memfd path: sendmsg(3) needs a file descriptor so we will have to connect to the socket path somehow (either opening the file or using the `UnixDatagram::connect` API.

Reply:
I think the `msghdr->msg_name` field (and `msglen`) can be used to tell `sendmsg()` about the recipient in the unconnected case.

Sidenote: I have no idea what's the result in case the specified address does not exist. Perhaps a nice experiment to try.

Reply:
Answering to my doubt above: `ENOENT` is returned when the address does not exists.